File: leetcode/easy/Q141_LinkedListCycle.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head:
            return False

        dict = {head: 0}
        node = head.next
        last = 0
        while node not in dict and node is not None:
            last += 1
            dict[node] = last
            node = node.next

        if node is None:
            return False
        else:
            return True

File: src/easy/Q141_LinkedListCycle.java
package easy;

import java.util.HashSet;

public class Q141_LinkedListCycle {
    class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
          val = x;
          next = null;
        }
    }
    public class Solution {
        public boolean hasCycle(ListNode head) {
            HashSet<ListNode> nodeSet = new HashSet<>();
            while (head != null){
                if(nodeSet.contains(head)){
                    return true;
                }
                nodeSet.add(head);
                head = head.next;
            }
            return false;
        }
    }
}
